File: native/ios/FireballNative/Core/ListenBrainzFeedParser.swift
import Foundation

enum ListenBrainzFeedParser {
    static func parseRecentListens(_ data: Data) -> [Track] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let payload = root["payload"] as? [String: Any],
              let listens = payload["listens"] as? [[String: Any]] else {
            return []
        }
        return listens.compactMap(parseRecentEntry)
    }

    static func parseTopRecordings(_ data: Data) -> [Track] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let payload = root["payload"] as? [String: Any],
              let recordings = payload["recordings"] as? [[String: Any]] else {
            return []
        }
        return recordings.compactMap(parseTopEntry)
    }

    private static func parseRecentEntry(_ listen: [String: Any]) -> Track? {
        guard let meta = listen["track_metadata"] as? [String: Any] else { return nil }
        let title = (meta["track_name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let artist = (meta["artist_name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !title.isEmpty, !artist.isEmpty else { return nil }
        return Track(
            id: "\(artist)::\(title)",
            videoId: nil,
            title: title,
            artist: artist,
            artwork: nil,
            url: nil,
            duration: nil,
            album: meta["release_name"] as? String,
            year: nil
        )
    }

    private static func parseTopEntry(_ rec: [String: Any]) -> Track? {
        let title = (rec["track_name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let artist = (rec["artist_name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !title.isEmpty, !artist.isEmpty else { return nil }
        return Track(
            id: "\(artist)::\(title)",
            videoId: nil,
            title: title,
            artist: artist,
            artwork: nil,
            url: nil,
            duration: nil,
            album: rec["release_name"] as? String,
            year: nil
        )
    }
}

File: native/android/core-data/src/main/kotlin/com/fireball/nativeapp/core/data/ListenBrainzFeedParser.kt
package com.fireball.nativeapp.core.data

import com.fireball.nativeapp.core.model.Track
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

object ListenBrainzFeedParser {
    private val json = Json { ignoreUnknownKeys = true }

    fun parseRecentListens(body: String): List<Track> {
        val root = runCatching { json.parseToJsonElement(body).jsonObject }.getOrNull() ?: return emptyList()
        val listens = root["payload"]?.jsonObject?.get("listens")?.jsonArray ?: return emptyList()
        return listens.mapNotNull { parseRecentEntry(it.jsonObject) }
    }

    fun parseTopRecordings(body: String): List<Track> {
        val root = runCatching { json.parseToJsonElement(body).jsonObject }.getOrNull() ?: return emptyList()
        val recordings = root["payload"]?.jsonObject?.get("recordings")?.jsonArray ?: return emptyList()
        return recordings.mapNotNull { parseTopEntry(it.jsonObject) }
    }

    private fun parseRecentEntry(listen: JsonObject): Track? {
        val meta = listen["track_metadata"]?.jsonObject ?: return null
        val title = meta["track_name"]?.jsonPrimitive?.content?.trim().orEmpty()
        val artist = meta["artist_name"]?.jsonPrimitive?.content?.trim().orEmpty()
        if (title.isBlank() || artist.isBlank()) return null
        val album = meta["release_name"]?.jsonPrimitive?.content
        return Track(
            id = "$artist::$title",
            title = title,
            artist = artist,
            album = album,
        )
    }

    private fun parseTopEntry(rec: JsonObject): Track? {
        val title = rec["track_name"]?.jsonPrimitive?.content?.trim().orEmpty()
        val artist = rec["artist_name"]?.jsonPrimitive?.content?.trim().orEmpty()
        if (title.isBlank() || artist.isBlank()) return null
        return Track(
            id = "$artist::$title",
            title = title,
            artist = artist,
            album = rec["release_name"]?.jsonPrimitive?.content,
        )
    }
}
